### PR TITLE
feat: Add exclusions and exclusion categories to gunmods

### PR DESCRIPTION
**Pull request by KheirFerrum**
## Purpose of change

- Expansion of #2541 

Allow further fine-tuning of what guns can and cannot be installed with specific gunmods, both for further development and specificity of vanilla gunmods and for the use of mod authors.

## Describe the solution

Adds `mod_exclusions` and `mod_exclusion_category` as optional fields in gunmods. This allows contributors to add specific blacklists (say you want to make a gun mod installable on all rifles _except_ the Rivtech battlerifle for instance), and blacklist categories (all rifles unless they are energy weapons).

Blacklists take priority over target categories _but not target ids_. Priority for inclusion/exclusion goes `mod_targets` > `mod_exclusions` > `mod_exclusion_category` > `mod_target_category`

Note that it _will_ scream at you if you try to include and exclude the same id or the same set of categories.

Changes the gyroscopic stabilizing system as a proof of concept for this system, and to ensure it's used _somewhere_ in the vanilla game to avoid it being lost to time.

## Testing

- Spawn an A7 laser rifle, a gyroscopic stabilizer and a modified gyroscopic stabilizer.
 - [x] Check that modified gyroscopic stabilizer is installable but gyroscopic stabilizer is not
 - Alter entry of gyroscopic stabilizer so it has `mod_targets` and `mod_exclusions` for `laser_rifle`, and extend `mod_target_category` to include `ENERGY_WEAPONS`
   - [x] Check that error message screams at you for doing this twice (once for `mod_targets` and `mod_exclusions` and once for `mod_target_category` and `mod_exclusion_category`)
   - [x] Check that you can install the gyroscopic stabilizer now.
 - [x] Check that if you supply invalid ids to `mod_exclusion` and `mod_exclusion_category` it screams at you. 

## Additional context

Not ideal to be using gyroscopic stabilizers as the proof of concept, but I can't think of anything that I can even remotely justify as an example right now. Maybe someone can look over the gunmods to see if anything nonsensical jumps out.

## Checklist

- [x] Document the changes in the appropriate location in the `doc/` folder.
- [x] If documentation for this feature does not exist, please write it or at least note its lack in PR description.
- [ ] New localizable fields need to be added to the `lang/bn_extract_json_strings.sh` script if it does not support them yet. (?)
- [x] If applicable, add checks on game load that would validate the loaded data.

### CODE REVIEW

**Review by scarf005 — COMMENTED** [review]
## additional screenshots

image
image
image

File: src/item.cpp [hunk 1]
```diff
@@ -7799,10 +7826,31 @@ ret_val<bool> item::is_gunmod_compatible( const item &mod ) const
         return ret_val<bool>::make_failure( _( "doesn't have enough room for another %s mod" ),
                                             mod.type->gunmod->location.name() );
 
-    } else if( !g_mod.usable.empty() || !g_mod.usable_category.empty() ) {
+    } else if( !g_mod.usable.empty() || !g_mod.usable_category.empty() || !g_mod.exclusion.empty() ||
+               !g_mod.exclusion_category.empty() ) {
+        // First check that it's not explicitly excluded by id.
+        bool unusable = g_mod.exclusion.count( this->typeId() );
```
Comment (scarf005) [inline src/item.cpp:7832]: might make sense to rename it to `excluded`.

Comment (KheirFerrum) [reply src/item.cpp:7832]: Sorry, how so?

Comment (scarf005) [reply src/item.cpp:7832]: bexause this line is highly confusing?

```c++
            if( usable || unusable ) {
```

Comment (KheirFerrum) [reply src/item.cpp:7832]: Ah, I thought you were talking about "exclusion".

**Review by scarf005 — APPROVED** [review]

**Pull request merged**